### ISSUE TITLE
Ordered beta regression documentation

### ISSUE THREAD

**Issue by saudiwin** [glmmTMB/glmmTMB#884]
Hi @bbolker - 

First of all, thanks so much for implementing my model in your package! I didn't know the model would work in an MLE framework, and wow, it is *fast*. Just a few thoughts on using the function: 

1. Is there a way to alert the user to the fact that ordered beta has cutpoints? I saw in another issue that the cutpoints are in the `psi` option, but that seems non-intuitive for most people. Maybe add a note in the family documentation to that effect?

2. You might want to consider including normalization of the outcome, perhaps by creating a `true_bounds` parameter to the `ordbeta` family. I use the following code in my package (feel free to use it yourself if it would save time):

```
  if(!is.null(true_bounds)) {

    min_out <- true_bounds[1]
    max_out <- true_bounds[2]
  } else {

    min_out <- min(outcome, na.rm=T)
    max_out <- max(outcome, na.rm=T)

    print(paste0("Normalizing using the observed bounds of ",min_out, " - ",
                 max_out,". If these are incorrect, please pass the bounds to use to the true_bounds parameter."))

  }

  trans_out <- (outcome - min_out) / (max_out - min_out)
```

The reason this is helpful is because some outcomes may not have observations at the bounds, such as a proportion where there are 0s but no 1s in the sample. Creating an explicit `true_bounds` parameter helps the user specify this situation and prevents false normalization (i.e. using the observed upper limit). Including normalization internally is also helpful for users who otherwise have to google and figure it out. 

Again, though, thanks for putting in the work on this! It is terrific and I will be using it for my own work as well 😄

**Comment by bbolker** [comment]
- Adding docs about cutpoints makes perfect sense
- I don't think we do any normalization at all, i.e. we assume the range is 0-1. I personally prefer that normalization be left to the user as a pre-processing step rather than being done automatically. Also, passing additional information (beyond the parameters and the standard info for mixed models, i.e. X and Z matrices and the response vector) to the C++ side of `glmmTMB` is something that requires us to extend the API a little bit (i.e. it requires more machinery that we could write, but haven't yet), which I wanted to avoid.

But the difference should definitely be documented.

**Comment by bbolker** [comment]
PS I see from the code in your comment that we could also do the normalization on the R side. Nevertheless, I think for now I'd prefer to document & leave it up to the user (but I'm open to further discussion with you and/or other glmmTMB devs (@mebrooks ?)

**Comment by saudiwin** [comment]
Obviously your call on this -- to give some context, I added this to my package when a user was confused about a variable that had bounds but those bounds weren't observed in the outcome. So let's say we have a percent, but the observed range is 0 - 89. The correct normalization is to use 0 as the lower bound and 100 as the upper bound, but if you aren't thinking, you'll end up normalizing with 89 as the upper bound. So the function and the `true_bounds` parameter is meant to avoid that mistake. It also saves time while being relatively little overhead as it can be done in R with pre-processing (maybe in the family function?).

In any case, great work on this! It's really fantastic!

**Comment by MToddASU** [comment]
A study I am working on has one of its primary outcome measured as proportions, where zeros and ones are both relatively abundant. So, the team is very interested in applying the ordinal beta regression method and curious to see how results from @saudiwin's ordregbeta and the glmmTMB implementation line up. 

So far, I've not been able to get glmmTMB going either with our project data or with the glmmTMB ordbeta examples here on GH. Every time,  no matter the data or the model, I get the following error:

Error in fitTMB(TMBStruc) : 
  negative log-likelihood is NaN at starting parameter values

For what it's worth, I am not having any issues getting ordregbeta to run with our data or the pew dataset used in the ordregbeta example. I know this doesn't give you much to go on, so if there is additional context that might be helpful, please let me know.

Thanks!

**Comment by bbolker** [comment]
Hmm. Can you confirm you're using glmmTMB version 1.1.5?  Do you have a data set you can share, either publicly (post a link here) or privately?

This adaptation of the base example in the `ordbetareg` package works for me (note you have to scale the `therm` variable yourself, as mentioned in the thread above), takes about 1 second.

```r
library(ordbetareg)
library(glmmTMB)
model_data <- dplyr::select(pew,therm,
                  education="F_EDUCCAT2_FINAL",
                  region="F_CREGION_FINAL",
                  income="F_INCOME_FINAL")
system.time(ord_fit <- glmmTMB(formula=therm/100 ~ education + income + (1|region),
         data=model_data,
         family = ordbeta))
```

**Comment by saudiwin** [comment]
Hi @MToddASU - there is an issue with the released version of `glmmTMB` and priors for cutpoints. See this code for instructions on how to run it in a way that is directly comparable to my package:

https://gist.github.com/saudiwin/791b3784e3c58ddcf7b839988a695ae2

**Comment by MToddASU** [comment]
@bbolker, yep, `glmmTMB` 1.1.5. I just tried the example you pasted in above and got the same message. I just tried the fix suggested by @saudiwin above, and that seemed to do the trick for both the example and my own data. 

With my own data, the solutions from `ordbetareg` and `glmmTMB` were largely similar, but with some noticeable differences in estimates for random effects and of the standard error for a fixed effect. There was an issue during the Bayesian estimation under `ordbetareg`\`brms` (divergent transition after warmup), which might explain some of the discrepancies, but I haven't had a chance to investigate further.

Thanks to you both for your efforts to make this approach available and accessible.

**Comment by bbolker** [comment]
That's weird, I could have sworn the fix was in the released version!  But I will double-check ...  You should definitely *not* ignore divergent transitions - in the Stan/HMC community (this is the machinery that `ordbetareg` is built on), the standard is that you should **never** accept a result where the sampling includes divergent transitions.

I would also expect that there *would* be differences between ordbetareg and glmmTMB results, due to some of the standard stuff:

* non-flat priors (ordbetareg) vs (implicitly) flat priors (on a particular scale) (glmmTMB)
* estimating posterior mean (ordbetareg) vs MLE (glmmTMB)
* integrating over latent effects by sampling (ordbetareg) vs Laplace approximation (glmmTMB)
* estimating SD by sampling (ordbetareg) vs Wald approximation (glmmTMB)
*

**Comment by nmasto** [comment]
Trying to implement a similar beta regression in `glmmTMB`. Just want to make sure I'm understanding thread correctly: `glmmTMB` currently does not have an option to specify starting values, or rather "bounds" for the beta regression? My response is bounded from -1 to 1. In such a case, did I understand correctly that in order to use the `ordbeta` family in `glmmTMB`, I'll need to standardize to 0-1 first (+0.001 to any zeros)?  Thanks for any clarification.

**Comment by bbolker** [comment]
You do need to standardize (I just added a note to the documentation in 4a75b12). You do *not* need to avoid the lower boundary, i.e. if your response `y` is in [-1, 1] you should scale your response to `(y-(-1))/(1-(-1)) = (y+1)/2`. (You can do this on the fly on the LHS of the formula if you want, i.e. `(y+1)/2 ~ ...`, although it might be better practice to do the transformation upstream.)

**Comment by nmasto** [comment]
Thanks @bbolker for quick response. Worked like a charm and _super_ fast compared to `brms` (which also required scaling).
